Index: zfs-linux/tests/zfs-tests/tests/functional/cli_root/zfs_program/zfs_program_json.ksh
===================================================================
--- zfs-linux.orig/tests/zfs-tests/tests/functional/cli_root/zfs_program/zfs_program_json.ksh
+++ zfs-linux/tests/zfs-tests/tests/functional/cli_root/zfs_program/zfs_program_json.ksh
@@ -100,10 +100,10 @@ typeset -a pos_cmds_out=(
 # the same as the input and the --sort-keys option was added.  Detect when
 # --sort-keys is supported and apply the option to ensure the expected order.
 #
-if python -m json.tool --sort-keys <<< "{}"; then
-	JSON_TOOL_CMD="python -m json.tool --sort-keys"
+if python3 -m json.tool --sort-keys <<< "{}"; then
+	JSON_TOOL_CMD="python3 -m json.tool --sort-keys"
 else
-	JSON_TOOL_CMD="python -m json.tool"
+	JSON_TOOL_CMD="python3 -m json.tool"
 fi
 
 typeset -i cnt=0
Index: zfs-linux/tests/zfs-tests/tests/functional/rsend/send_encrypted_files.ksh
===================================================================
--- zfs-linux.orig/tests/zfs-tests/tests/functional/rsend/send_encrypted_files.ksh
+++ zfs-linux/tests/zfs-tests/tests/functional/rsend/send_encrypted_files.ksh
@@ -87,7 +87,7 @@ log_must xattrtest -f 10 -x 3 -s 32768 -
 # OpenZFS issue #7432
 log_must zfs set compression=on xattr=sa $TESTPOOL/$TESTFS2
 log_must touch /$TESTPOOL/$TESTFS2/attrs
-log_must eval "python -c 'print \"a\" * 4096' | \
+log_must eval "python3 -c 'print \"a\" * 4096' | \
 	set_xattr_stdin bigval /$TESTPOOL/$TESTFS2/attrs"
 log_must zfs set compression=off xattr=on $TESTPOOL/$TESTFS2
 
Index: zfs-linux/tests/zfs-tests/tests/functional/rsend/send_realloc_dnode_size.ksh
===================================================================
--- zfs-linux.orig/tests/zfs-tests/tests/functional/rsend/send_realloc_dnode_size.ksh
+++ zfs-linux/tests/zfs-tests/tests/functional/rsend/send_realloc_dnode_size.ksh
@@ -88,7 +88,7 @@ log_must zfs snapshot $POOL/fs@c
 # 4. Create an empty file and add xattrs to it to exercise reclaiming a
 #    dnode that requires more than 1 slot for its bonus buffer (Zol #7433)
 log_must zfs set compression=on xattr=sa $POOL/fs
-log_must eval "python -c 'print \"a\" * 512' |
+log_must eval "python3 -c 'print \"a\" * 512' |
     set_xattr_stdin bigval /$POOL/fs/attrs"
 log_must zfs snapshot $POOL/fs@d
 
