Index: zfs-linux-0.8.3/module/Makefile.in
===================================================================
--- zfs-linux-0.8.3.orig/module/Makefile.in
+++ zfs-linux-0.8.3/module/Makefile.in
@@ -1,11 +1,11 @@
-obj-m += avl/
-obj-m += icp/
-obj-m += lua/
-obj-m += nvpair/
-obj-m += spl/
-obj-m += unicode/
-obj-m += zcommon/
-obj-m += zfs/
+obj-y += avl/
+obj-y += icp/
+obj-y += lua/
+obj-y += nvpair/
+obj-y += spl/
+obj-y += unicode/
+obj-y += zcommon/
+obj-y += zfs/
 
 INSTALL_MOD_DIR ?= extra
 
@@ -60,12 +60,12 @@ modules_install:
 modules_uninstall:
 	@# Uninstall the kernel modules
 	kmoddir=$(DESTDIR)$(INSTALL_MOD_PATH)/lib/modules/@LINUX_VERSION@
-	list='$(obj-m)'; for objdir in $$list; do \
+	list='$(obj-y)'; for objdir in $$list; do \
 		$(RM) -R $$kmoddir/$(INSTALL_MOD_DIR)/$$objdir; \
 	done
 
 distdir:
-	list='$(obj-m)'; for objdir in $$list; do \
+	list='$(obj-y)'; for objdir in $$list; do \
 		(cd @top_srcdir@/module && find $$objdir \
 		-name '*.c' -o -name '*.h' -o -name '*.S' | \
 		xargs -r cp --parents -t @abs_top_builddir@/module/$$distdir); \
