Description: Fix cross-compile of the dkms module.
Author: Dimitri John Ledkov <dimitri.ledkov@canonical.com>


Index: zfs-linux/scripts/dkms.mkconf
===================================================================
--- zfs-linux.orig/scripts/dkms.mkconf
+++ zfs-linux/scripts/dkms.mkconf
@@ -50,6 +50,12 @@ PRE_BUILD="configure
     }
   )
   \$(
+    [[ -n \"\${CROSS_COMPILE}\" ]] && \\
+    {
+      echo --host=\"\${CROSS_COMPILE%-}\"
+    }
+  )
+  \$(
     [[ -r \${PACKAGE_CONFIG} ]] \\
     && source \${PACKAGE_CONFIG} \\
     && shopt -q -s extglob \\
